### ISSUE TITLE
Observation: Input Probabilities for Viterbi Decoder Not Log-Transformed

### ISSUE THREAD

**Issue by grmos** [BernhoferM/TMbed#13]
Hello, Michael!
I was reviewing the implementation of the Viterbi decoder in the project and observed that the input probabilities are passed directly in their linear form (as produced by the softmax function). From my understanding, this particular Viterbi implementation involves summing emissions and transitions (which are on a logarithmic scale) instead of multiplying them, which suggests that log-transformed probabilities might be a better fit.

Suggested Adjustment
`log_probabilities = torch.log(probabilities)`
`prediction = decoder(log_probabilities, mask).byte()`

Best regards,
Grigoris Moschos

**Comment by BernhoferM** [comment]
Hi.

You are correct: the decoder is looking for the path with the highest sum of probabilities (as noted in the SOM of the paper).

However, this is intentional as it is a bit less sensitive to single outliers; a single very low probability can completely block a path when using log-probabilities (i.e., searching for the path of highest probability product). I *think* to remember that I tested both versions and then decided on the sum instead of the product... but it's been a while and I'm not 100% sure. If I have the time, I might test and validate it again to make sure.

If you want to change it to the probability product, then your changes are correct. Though I would probably add a small epsilon (e.g., 1e-9) to the probabilities before taking the log (for numerical stability). Or you could directly change the [Softmax](https://github.com/BernhoferM/TMbed/blob/main/tmbed/tmbed.py#L129) calculation to LogSoftmax. Though then the averaging of probabilities (over the five models) changes from arithmetic mean to geometric mean.

Regards
Michael

**Comment by grmos** [comment]
Hi Michael,

Thank you for your detailed response and for explaining the rationale behind using summation instead of multiplication. However, I think there might be a misunderstanding regarding my concern. My question isn't about whether summing probabilities is a better approach compared to multiplying them, but rather about the scale consistency of the values being summed.

From what I understand, the transition probabilities are already in logarithmic form(initiali values 0,-100), whereas the emission probabilities are in their linear form (raw probabilities output by the softmax function). When summing these two, they are not on the same scale. Typically, in Viterbi decoding, both transition and emission probabilities are either converted to the logarithmic scale or left in their original form, but they should match for the summation to be meaningful.

**Comment by BernhoferM** [comment]
Ah, I see. Yes, the initial values (0, -100) can be seen as log-space values (corresponding to 1.0 and ~0.0). However, in the context of summing up probabilities (which are in [0.0, 1.0]) they rather act as a neutral value (0) and an extreme penalty (-100) to essentially block any undesirable paths (e.g., directly going from H to B). There would need to be a path that has a probability sum 100 higher than any alternative path to justify a single one of those transitions (which I haven't seen in any of my test cases).

However, this only works as there are no other transitions besides "yes" (0) and "no" (-100). If you wanted to use different transition scores for each of the possible state transitions, then you'd need to make sure that the value ranges make sense regarding either log or probability value-space.

Note that for TMbed the transition scores are not learned, i.e., changed after the initialization. The decoder is essentially just used to find the best path based on the predicted state probabilities, only knowing about allowed and disallowed transitions (which define a rather simple grammar). In short, the Viterbi decoding is highly simplified.

I hope this answered your questions. Let me know if not.  :)

Regards
Michael
